File: codebases/backend/api/upload/upload_file_handler.php
<?php

require_once Env::get('BACKEND_DTO_RESPONSE_PATH') . '/php/FileUploadResponse.php';
require_once Env::get('BACKEND_DTO_RESPONSE_PATH') . '/php/ApiResponse.php';
require_once Env::get('BACKEND_DTO_DETAIL_PATH') . '/php/FileUploadDetail.php';
require_once Env::get('BACKEND_DTO_DETAIL_PATH') . '/php/ExtractResult.php';

// JSON 파일에서 상수 읽기
$jsonString = file_get_contents(Env::get('BACKEND_CONSTANTS_PATH') . '/FileType.json');
$fileType = json_decode($jsonString, true);
$jsonString = file_get_contents(Env::get('BACKEND_CONSTANTS_PATH') . '/FileHandleStatus.json');
$fileHandleStatus = json_decode($jsonString, true);

// Python 스크립트의 절대 경로
$scriptPath = Env::get('BACKEND_API_PATH').'/upload/upload_file_handler.py';
$extractScriptPath = Env::get('BACKEND_API_PATH').'/extract/extract_text.py';
$highlightScriptPath = Env::get('BACKEND_API_PATH').'/pdf_styling/highlight_pdf.py';
$textDirPath = Env::get('UPLOAD_PATH').'/texts/';
$uploadDir = Env::get('UPLOAD_PATH').'/pdfs/';
$uploadedFiles = [];
$responseData = [];

// 디렉토리가 존재하지 않으면 생성
if (!is_dir($uploadDir)) {
    if (!mkdir($uploadDir, 0770, true)) {
        error_log("Failed to create directory: $uploadDir");
    }
}

// Python 스크립트 실행을 위한 헬퍼 함수
function executePythonScript($scriptPath, $args) {
    $command = sprintf(
        'env %s %s %s %s',
        Env::getPythonEnvString(),
        escapeshellarg(Env::get('PYTHON_PATH')),
        escapeshellarg($scriptPath),
        escapeshellarg($args)
    );
    
    logGeneralError("Python 스크립트 실행: $command");
    $output = shell_exec($command . " 2>&1");
    
    if ($output === null) {
        logGeneralError("Python 스크립트 실행 실패: $command");
        throw new Exception("Python 스크립트 실행 실패");
    }
    
    logGeneralError("Python 스크립트 출력: " . $output);
    return $output;
}

try {
    // 업로드된 파일의 개수 로그 기록
    $fileCount = count($_FILES['files']['tmp_name']);
    logGeneralError("파일 업로드 처리 시작");
    
    foreach ($_FILES['files']['tmp_name'] as $key => $tmpName) {
        $fileName = basename($_FILES['files']['name'][$key]);
        $filePath = $uploadDir.$fileName;
        logGeneralError("!!파일 경로: $filePath");
        
        if (move_uploaded_file($tmpName, $filePath)) {
            // 1. 파일 유형 확인 스크립트 실행
            $output = executePythonScript($scriptPath, $filePath);
            $result = json_decode($output, true);
            
            if ($result === null) {
                throw new Exception("파이썬 스크립트 오류: " . json_last_error_msg());
            }
            
            // 파일 유형에 따라 처리
            if ($result['fileType'] === $fileType['FILE_TYPE_TEXT']) {
                $baseFileName = pathinfo($fileName, PATHINFO_FILENAME);
                $extractFilePath = $textDirPath.$baseFileName.".txt";
                
                // 2. 텍스트 추출 스크립트 실행
                $extractOutput = executePythonScript($extractScriptPath, $extractFilePath);
                $extractResult = json_decode($extractOutput, true);
                
                if ($extractResult['status'] === 'success') {
                    $extractResultObject = ExtractResult::fromArray($extractResult);
                    $extractResultObject->setFileName($fileName);
                    $extractResultObject->setFileType($fileType['FILE_TYPE_TEXT']);
                    $extractResults[] = $extractResultObject;
                } else {
                    $extractResults[] = new ExtractResult(
                        $fileName, 
                        $fileHandleStatus['STATUS_FAILURE'], 
                        $fileType['FILE_TYPE_TEXT'], 
                        null, null, null, null, null, null
                    );
                }
            }
            else if  ($fileType['FILE_TYPE_IMAGE'] === $result['fileType'] ) {
                $extractResults[] = new ExtractResult($fileName, $fileHandleStatus['STATUS_SUCCESS'], $fileType['FILE_TYPE_IMAGE'], null, null, null, null, null, null);
                // 이미지 파일인 경우

            }
            else {
                $error = error_get_last();
                logGeneralError($filePath, $error ? $error['message'] : '파일 유형 오류');
                $extractResults[] = new ExtractResult($fileName, $fileHandleStatus['STATUS_FAILURE'], $fileType['FILE_TYPE_UNKNOWN'], null, null, null, null, null, null);
            }
        } 
        else {
            $error = error_get_last();
            logGeneralError($filePath, $error ? $error['message'] : '파일 업로드 실패');
            $extractResults[] = new ExtractResult($fileName, $fileHandleStatus['STATUS_FAILURE'], $fileType['FILE_TYPE_UNKNOWN'], null, null, null, null, null, null);
        }
    }
   
    $response = new FileUploadResponse('success', '파일 업로드 및 처리 완료', $extractResults);
    echo $response->toJson();
    
    // 3. 하이라이트 작업 실행
    logGeneralError("하이라이트 작업 시작");
    highlight_text_on_specific_pages($response->getExtractResults(), $highlightScriptPath);
    
} catch (Exception $e) {
    $last_error = error_get_last();
    logGeneralError("파일 처리 중 예외 발생: " . $e->getMessage());
    $response = new ApiResponse('error', '파일 처리 중 오류 발생: ' . $e->getMessage() . " " . $last_error['message']);
    echo $response->toJson();
    exit;
}

function highlight_text_on_specific_pages($extractResults, $highlightScriptPath) {
    // JSON 문자열을 인코딩할 때 옵션을 조정
    $extractResultsJson = json_encode($extractResults, JSON_UNESCAPED_UNICODE);
    
    try {
        // 4. 하이라이트 스크립트 실행
        $output = executePythonScript($highlightScriptPath, $extractResultsJson);
        if ($output === null) {
            logGeneralError("하이라이트 작업 실패");
        } else {
            logGeneralError("하이라이트 작업 결과: " . $output);
        }
    } catch (Exception $e) {
        logGeneralError("하이라이트 작업 중 예외 발생: " . $e->getMessage());
    }
}
?>

File: codebases/backend/api/upload/upload_file_handler.py
import pdfplumber
import json
import sys
import os
import logging
sys.path.append(os.path.abspath('/var/www/html/invoiceProject/codebases'))
from backend.DTO.detail.py.FileUploadDetail import FileUploadDetail
from backend.config.logging_config import setup_logging

# 로깅 설정 초기화
setup_logging()

# JSON 파일에서 상수 읽기
try:
    with open('/var/www/html/invoiceProject/codebases/backend/constants/FileType.json', 'r') as file:
        file_type = json.load(file)
except Exception as e:
    logging.error("FileType.json 파일을 읽는 중 오류 발생", exc_info=True)

try:
    with open('/var/www/html/invoiceProject/codebases/backend/constants/FileHandleStatus.json', 'r') as file:
        file_handle_status = json.load(file)
except Exception as e:
    logging.error("FileHandleStatus.json 파일을 읽는 중 오류 발생", exc_info=True)

def process_pdf(file_path):
    try:
        with pdfplumber.open(file_path) as pdf:
            text_found = False
            output_folder = "/var/www/html/invoiceProject/uploads/texts"
            os.makedirs(output_folder, exist_ok=True)
            
            # PDF 파일명에서 확장자 제거
            base_filename = os.path.splitext(os.path.basename(file_path))[0]
            text_file_path = os.path.join(output_folder, f"{base_filename}.txt")
            
            with open(text_file_path, "w", encoding="utf-8") as text_file:
                for page_number, page in enumerate(pdf.pages):
                    text = page.extract_text(layout=True)
                    if text:
                        text_found = True
                        text_file.write(f"Page {page_number + 1}:\n{text}\n\n")
            
            if text_found:
                logging.debug(f"텍스트가 포함된 PDF 처리 성공: {file_path}")
                return FileUploadDetail(status=file_handle_status['STATUS_SUCCESS'], file_type=file_type['FILE_TYPE_TEXT']).to_dict()
            else:
                logging.warning(f"이미지로만 구성된 PDF: {file_path}")
                return FileUploadDetail(status=file_handle_status['STATUS_SUCCESS'], file_type=file_type['FILE_TYPE_IMAGE']).to_dict()

    except FileNotFoundError:
        logging.error(f"파일을 찾을 수 없습니다: {file_path}", exc_info=True)
        return FileUploadDetail(status=file_handle_status['STATUS_ERROR'], file_type=file_type['FILE_TYPE_UNKNOWN']).to_dict()

    except Exception as e:
        logging.error(f"PDF 처리 중 오류 발생: {file_path}", exc_info=True)
        return FileUploadDetail(status=file_handle_status['STATUS_ERROR'], file_type=file_type['FILE_TYPE_UNKNOWN']).to_dict()

if __name__ == '__main__':
    logging.debug("upload_file_handler.py 실행")
    if len(sys.argv) > 1:
        file_path = sys.argv[1]
        result = process_pdf(file_path)
        print(json.dumps(result))
    else:
        logging.error("파일 경로가 제공되지 않았습니다.")
        result = FileUploadDetail(status=file_handle_status['STATUS_ERROR'], file_type=file_type['FILE_TYPE_UNKNOWN']).to_dict()
        print(json.dumps(result))